diff --git syntax/quote.go syntax/quote.go
index 741e106..ada27a9 100644
--- syntax/quote.go
+++ syntax/quote.go
@@ -156,6 +156,10 @@ func unquote(quoted string) (s string, triple, isByte bool, err error) {
 				quoted = quoted[1:]
 			}
 			if !isByte && n > 127 {
+				if n < 256 {
+					buf.WriteRune(rune(n))
+					break
+				}
 				err = fmt.Errorf(`non-ASCII octal escape \%o (use \u%04X for the UTF-8 encoding of U+%04X)`, n, n, n)
 				return
 			}
